### PR TITLE
Updated ADC Implementation to support ESP-IDF 5.0

### PR DESCRIPTION
**Pull request by SuperChamp234**
I am issuing this PR to update our ADC Implementation for ESP-IDF 5.0.

As per pull request #87 , I have rewritten the ADC implementation to use methods from the new ESP-IDF 5.0 library.

It makes use of the `esp_adc` component instead of the depreciated `esp_adc_cal`.

The new ADC methods require the use of an `adc_handle_t` when trying to access the ADC object and perform actions.

Please review and let me know the changes, thank you.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -76,8 +79,8 @@ dkms.conf
...
-!.vscode/tasks.json
-!.vscode/launch.json
-!.vscode/extensions.json
-*.code-workspace
\ No newline at end of file
+.vscode/settings.json
+.vscode/tasks.json
+.vscode/launch.json
+.vscode/extensions.json
+*.code-workspace
```
Comment (Krishna-13-cyber) [inline .gitignore:86]: Add newline to cancle this diff, aside from that LGTM :)

File: CMakeLists.txt [hunk 2]
```diff
@@ -10,5 +12,5 @@ endif()
 
 idf_component_register(SRCS "${srcs}"
                        INCLUDE_DIRS "${include_dirs}"
-                       REQUIRES driver efuse esp32 esp_adc_cal esp_common esp_event 
-                                esp_rom freertos heap log soc spiffs)
\ No newline at end of file
+                       REQUIRES driver efuse esp_adc esp_common esp_event 
+                                esp_rom esp_timer freertos heap log soc spiffs)
```
Comment (Krishna-13-cyber) [inline CMakeLists.txt:16]: LGTM

File: include/adc.h [hunk 3]
```diff
@@ -39,36 +41,25 @@
...
-#define DEFAULT_VREF 1100
+#define ADC_ATTEN ADC_ATTEN_DB_11
 
-/**
- * @brief Configure width and attenuation of ADC 1
- * 
- * @return esp_err_t - returns ESP_OK if 11dB attenuation and 12 bit width configured to ADC 1, else it returns ESP_ERR_INVALID_ARG.  
- **/
-esp_err_t config_adc1();
+//define a struct that holds the adc1 handle and calibration handles for five channels in an array
```
Comment (Krishna-13-cyber) [inline include/adc.h:46]: Use multiline comments for documenting this.

File: include/adc.h [hunk 4]
```diff
@@ -39,36 +41,25 @@
...
 
-/**
- * @brief Read raw adc value of given adc pin.
- * 
- * @param adc_pin One of the GPIO pin to which LSA is connected (36/39/35/34)
- * 
- * @return esp_err_t -returns raw reading of adc pin if lsa pin is passed to function, else it returns ESP_ERR_INVALID_ARG.
- **/
-int read_adc(int adc_pin);
+int read_adc(adc_handle_t adc_handle, int gpio);
```
Comment (Krishna-13-cyber) [inline include/adc.h:63]: Add documentation for this function.

File: include/lsa.h [hunk 5]
```diff
@@ -62,6 +62,6 @@ esp_err_t enable_line_sensor();
  * @return Returns a pointer to a struct that contains multisampled adc readings 
            for all the pins specified in lsa_pins array
 **/
-line_sensor_array read_line_sensor();
+line_sensor_array read_line_sensor(adc_handle_t adc_handle);
```
Comment (Krishna-13-cyber) [inline include/lsa.h:66]: LGTM

File: include/sra_board.h [hunk 6]
```diff
@@ -39,4 +39,4 @@
 #include "oled.h"
 #endif
 
-#endif
\ No newline at end of file
+#endif
```
Comment (Krishna-13-cyber) [inline include/sra_board.h:42]: Add a new line :)

File: src/adc.c [hunk 7]
```diff
@@ -24,82 +24,143 @@
...
-    CHECK(adc1_config_channel_atten(ADC_CHANNEL_4, ADC_ATTEN_DB_11));
-    CHECK(adc1_config_channel_atten(ADC_CHANNEL_7, ADC_ATTEN_DB_11));
-    CHECK(adc1_config_channel_atten(ADC_CHANNEL_6, ADC_ATTEN_DB_11));
-    CHECK(adc1_config_channel_atten(ADC_CHANNEL_3, ADC_ATTEN_DB_11));
-    CHECK(adc1_config_channel_atten(ADC_CHANNEL_0, ADC_ATTEN_DB_11));
-
-    ESP_LOGI(TAG_ADC, "Configured ADC_1 to 12 Bit and 11dB attenuation");
+    adc_oneshot_chan_cfg_t config = {
+        .bitwidth = ADC_BITWIDTH_12,
+        .atten = ADC_ATTEN,
```
Comment (Krishna-13-cyber) [inline src/adc.c:34]: You can use `ADC_ATTEN_DB_11` directly here

File: src/adc.c [hunk 8]
```diff
@@ -24,82 +24,143 @@
...
+    };
+    ESP_LOGI(TAG, "ADC1 handle address: %p", adc_obj->adc1_handle);
```
Comment (Krishna-13-cyber) [inline src/adc.c:36]: Why do we need to log the address of the adc1 handle? If not necessary remove it.

File: src/adc.c [hunk 9]
```diff
@@ -24,82 +24,143 @@
...
+
+esp_err_t calib_init(adc_obj_t *adc_obj)
+{
+    for (int i = 0; i < sizeof(adc_io) / sizeof(adc_io[0]); i++)
     {
-        ESP_LOGI(TAG_ADC, "Characterized using Default Vref");
+        adc_channel_t channel;
+        adc_unit_t unit;
+        ESP_ERROR_CHECK(adc_oneshot_io_to_channel(adc_io[i], &unit, &channel));
+        adc_obj->do_calib[i] = adc_calibration_init(ADC_UNIT_1, channel, ADC_ATTEN, &adc_obj->adc1_cali_handle[i]);
```
Comment (Krishna-13-cyber) [inline src/adc.c:105]: You can use `ADC_ATTEN_DB_11` directly here

File: src/adc.c [hunk 10]
```diff
@@ -24,82 +24,143 @@
...
-
-    ESP_LOGI(TAG_ADC, "Configured and Characterized adc 1");
+    adc_obj_t *ret = NULL;
+    ret = calloc(1, sizeof(adc_obj_t));
+    ESP_ERROR_CHECK(ret == NULL ? ESP_ERR_NO_MEM : ESP_OK);
+    adc_oneshot_unit_init_cfg_t init_config1 = {
+        .unit_id = ADC_UNIT_1,
+    };
+    ESP_ERROR_CHECK(adc_oneshot_new_unit(&init_config1, &ret->adc1_handle));
+    ESP_LOGI(TAG, "Handle address: %p", ret->adc1_handle);
```
Comment (Krishna-13-cyber) [inline src/adc.c:119]: Why do we need to log the address of the adc handle? If not necessary remove it.

File: src/adc.c [hunk 11]
```diff
@@ -24,82 +24,143 @@
...
+    config_adc1(ret);
+    calib_init(ret);
+    //assign return value to adc_handle
```
Comment (Krishna-13-cyber) [inline src/adc.c:122]: Remove this comment as it is apparent from the code.

File: src/i2cdev.c [hunk 12]
```diff
@@ -232,7 +232,7 @@ static esp_err_t i2c_setup_port(const i2c_dev_t *dev)
     uint32_t ticks = dev->timeout_ticks ? dev->timeout_ticks : I2CDEV_MAX_STRETCH_TIME;
     if ((ticks != t) && (res = i2c_set_timeout(dev->port, ticks)) != ESP_OK)
         return res;
-    ESP_LOGD(TAG, "Timeout: ticks = %d (%d usec) on port %d", dev->timeout_ticks, dev->timeout_ticks / 80, dev->port);
+    ESP_LOGD(TAG, "Timeout: ticks = %d (%d usec) on port %d", (int)dev->timeout_ticks, (int)(dev->timeout_ticks / 80), (int)dev->port);
 #endif
```
Comment (Krishna-13-cyber) [inline src/i2cdev.c:237]: LGTM

File: src/lsa.c [hunk 13]
```diff
@@ -45,7 +45,7 @@ line_sensor_array read_line_sensor()
     {
         for (int j = 0; j < 5; j++)
         {
-            line_sensor_readings.adc_reading[j] = line_sensor_readings.adc_reading[j] + read_adc(line_sensor_pins[j]);
+            line_sensor_readings.adc_reading[j] = line_sensor_readings.adc_reading[j] + read_adc(adc_handle, line_sensor_pins[j]);
         }
     }
```
Comment (Krishna-13-cyber) [inline src/lsa.c:51]: LGTM

File: CMakeLists.txt (hunk 2 above)
Comment (RISHI27-dot) [inline CMakeLists.txt:16]: LGTM

File: include/adc.h [hunk 14]
```diff
@@ -27,9 +27,11 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
```
Comment (RISHI27-dot) [inline include/adc.h:30]: Why do we require this?

File: include/adc.h (hunk 3 above)
Comment (RISHI27-dot) [inline include/adc.h:46]: Add documentation for this using multiline comments.

File: include/adc.h (hunk 4 above)
Comment (RISHI27-dot) [inline include/adc.h:63]: Update documentation for this function using multiline comments.

File: src/adc.c (hunk 8 above)
Comment (RISHI27-dot) [inline src/adc.c:36]: Why do we need to log the address of the handle here? If not necessary remove it.

File: src/adc.c (hunk 7 above)
Comment (RISHI27-dot) [inline src/adc.c:34]: You can use `ADC_ATTEN_DB_11` here directly.

File: src/adc.c (hunk 9 above)
Comment (RISHI27-dot) [inline src/adc.c:105]: You can use `ADC_ATTEN_DB_11` here directly.

File: src/adc.c (hunk 10 above)
Comment (RISHI27-dot) [inline src/adc.c:119]: Why do we need to log the address of the handle here? If not necessary remove it.

File: src/adc.c (hunk 11 above)
Comment (RISHI27-dot) [inline src/adc.c:122]: Remove this comment it is apparent from the code.

File: src/lsa.c (hunk 13 above)
Comment (RISHI27-dot) [inline src/lsa.c:51]: LGTM.

File: include/sra_board.h (hunk 6 above)
Comment (RISHI27-dot) [inline include/sra_board.h:42]: check for the diffs resulting from unnecessary addition and removal of new lines here and in all instances above this point.

File: src/adc.c (hunk 7 above)
Comment (SuperChamp234) [reply src/adc.c:34]: One might want to change the calibration parameters, so having it changed in one place might be convenient.

File: src/adc.c (hunk 9 above)
Comment (SuperChamp234) [reply src/adc.c:105]: One might want to change the calibration parameters, so having it changed in one place might be convenient.

File: include/adc.h (hunk 14 above)
Comment (SuperChamp234) [reply include/adc.h:30]: It is not necessary, I will remove it

File: include/adc.h (hunk 3 above)
Comment (SuperChamp234) [reply include/adc.h:46]: Noted.

File: include/adc.h (hunk 4 above)
Comment (SuperChamp234) [reply include/adc.h:63]: Noted.

File: src/adc.c (hunk 8 above)
Comment (SuperChamp234) [reply src/adc.c:36]: Yes, will remove.

File: src/adc.c (hunk 10 above)
Comment (SuperChamp234) [reply src/adc.c:119]: Yes, will remove.

File: src/adc.c (hunk 11 above)
Comment (SuperChamp234) [reply src/adc.c:122]: Noted.

File: include/sra_board.h (hunk 6 above)
Comment (SuperChamp234) [reply include/sra_board.h:42]: Noted.

**Comment by SuperChamp234** [comment]
Hi @RISHI27-dot ,
I have made the changes as requested, please review.

File: src/adc.c (hunk 7 above)
Comment (RISHI27-dot) [reply src/adc.c:34]: Any specific reason for not doing the same thing with `ADC_BITWIDTH` ?

Comment (SuperChamp234) [reply src/adc.c:34]: I missed it, will also create a definition for the bitwidth.

File: src/i2cdev.c [hunk 15]
```diff
@@ -222,7 +222,7 @@ static esp_err_t i2c_setup_port(const i2c_dev_t *dev)
         states[dev->port].installed = true;
 
         memcpy(&states[dev->port].config, &temp, sizeof(i2c_config_t));
-        ESP_LOGD(TAG, "I2C driver successfully reconfigured on port %d", dev->port);
+        ESP_LOGD(TAG, "I2C driver successfully reconfigured on port %d", (int)dev->port);
```
Comment (RISHI27-dot) [inline src/i2cdev.c:225]: typecasting is not necessary here.

File: src/adc.c (hunk 7 above)
Comment (SuperChamp234) [reply src/adc.c:34]: Done

File: include/adc.h
Comment (RISHI27-dot) [inline include/adc.h:1]: LGTM!!

**Review by RISHI27-dot — APPROVED** [review]

File: src/i2cdev.c (hunk 15 above)
Comment (SuperChamp234) [reply src/i2cdev.c]: Updated

**Pull request merged**